projects/project4/src/main/scala/net/sansa_stack/template/spark/rdf/Algorithm.scala

```diff
@@ -0,0 +1,527 @@
+          val scheme = sPrevious(i)
+          val newSchemes = generateBlockingSchemes(Array(scheme), optimalScheme, "and")
+
+          if (!newSchemes.isEmpty) {
+            val andScheme = newSchemes(0)
+
+            val fp = falsePositives(spark, scheme, trainingSet)
+
+            val andSchemeFP = falsePositives(spark, andScheme, trainingSet)
+
+            if ((optimumFP >= andSchemeFP) && fp >= andSchemeFP) {
+              S = S :+ andScheme
+            } else {
+              S = S :+ scheme
+            }
+          }
+        }
+
+      } else {
+
+        val optimalFN = falseNegatives(spark, optimalScheme, trainingSet)
+        val optimalTP = truePositives(spark, optimalScheme, trainingSet)
+
+        for (i <- 0 to sPrevious.length - 1) {
+          val scheme = sPrevious(i)
+          val fn = falseNegatives(spark, scheme, trainingSet)
+          val tp = truePositives(spark, scheme, trainingSet)
+
+          val newSchemes = generateBlockingSchemes(Array(scheme), optimalScheme, "or")
+
+          if (!newSchemes.isEmpty) {
+            val orScheme = newSchemes(0)
+            val orSchemeFN = falseNegatives(spark, orScheme, trainingSet)
+            val orSchemeTP = truePositives(spark, orScheme, trainingSet)
+
+            if (((optimalFN/optimalTP) >= (orSchemeFN/orSchemeTP)) && ((fn/tp) >= (orSchemeFN/orSchemeTP))) {
+              S = S :+ orScheme
+            } else {
+              S = S :+ scheme
+            }
+          }
+        }
+      }
+    }
+
+    println(s"Optimal Scheme ${optimalScheme}")
+    val reductionRatio = Experiments.reductionRatio(trainingSet, optimalScheme)
+    val completness = Experiments.pairCompleteness(trainingSet, optimalScheme)
+    val quality = Experiments.pairQuality(trainingSet, optimalScheme)
+    val fmeas = Experiments.fMeasure(completness, quality)
+    val cSatisfaction = Experiments.constraintSatisfaction(constraintSatisfactionNs, n)
+
+    println("*** Results ***")
+
+    println(s"Budget ${budget}")
+    println(s"Human Oracle Percentage ${humanOracleThresholdPercentage}")
+    println(s"Initial Sample Size ${sampleSize}")
+    println(s"Sample Size ${trainingSet.count()}")
+    println(s"Error Rate Threshold ${errorRateThreshold}")
+    println(s"Reduction Ratio ${reductionRatio}")
+    println(s"Pair Completeness ${completness}")
+    println(s"Pair Quality ${quality}")
+    println(s"F Measure ${fmeas}")
+    println(s"Constraint Satisfaction ${cSatisfaction}")
+
+    return optimalScheme
+  }
+
+  /** Calculates false positives given a training set dataframe and a blocking scheme.
+    *
+    *  @param spark the spark session object.
+    *  @param scheme blocking scheme.
+    *  @param trainingSet training set dataframe.
+    */
+  def falsePositives(spark: SparkSession, scheme: String, trainingSet: DataFrame): Long = {
+
+    val count = spark.sparkContext.longAccumulator
+    val columnNames = trainingSet.columns
+
+    trainingSet.foreach { row =>
```

Comment:
Is there a possibility of doing this using `flapMap` or we have to materialize it in order to get the header(schema)?. If so, it will keep the data to the workers and count the accumulated value instead, which is the way you have used to count the value on the workers and get the accumulated value in the end. But the same approach hasn't been used on the code as it brings all the data to the driver and there is no need to apply accumulator variable at all. 
Consider keeping the data there and compute just the value of the count (accumulator).

Reply:
Thanks for the feedback. You were right. We have shifted to use flatmap and just call count method on the result instead of using foreach.

Reply:
Cool!

Reply:
@GezimSejdiu You can also see the updated report in the latest commit. Thanks.